### PR TITLE
Hex preview

### PR DESCRIPTION
**Pull request by jstucke**
- added hex preview for files that are neither text files nor images
- added highlighting for hex output (xxd style)
- added intercom function for retrieving part of a file

### CODE REVIEW

**Comment by dorpvom** [comment]
```
Traceback (most recent call last):
  File "flask_app_wrapper.py", line 28, in <module>
    from web_interface.frontend_main import WebFrontEnd
  File "/home/dorp/git/FACT_core/src/./web_interface/frontend_main.py", line 7, in <module>
    from web_interface.components.ajax_routes import AjaxRoutes
  File "/home/dorp/git/FACT_core/src/./web_interface/components/ajax_routes.py", line 14, in <module>
    from web_interface.components.hex_highlighting import highlight_hex
ModuleNotFoundError: No module named 'web_interface.components.hex_highlighting'
```

**Comment by dorpvom** [comment]
Did we not have this functionality and deprecated it in the meantime?

**Comment by jstucke** [comment]
> Did we not have this functionality and deprecated it in the meantime?

Yes and no: We had a hex preview but it was not that well integrated and we gave it up in favor of the Radare GUI. I just felt that the Radare GUI is a bit much overhead for just taking a quick peek at the header or content. 

This preview comes with color highlighting and a variable offset and length. Also it is pretty efficient, since it only reads the part of the file that is actually displayed.

File: src/intercom/front_end_binding.py [hunk 1]
```diff
@@ -34,11 +34,15 @@ def get_available_analysis_plugins(self):
...
             plugin_dict = pickle.loads(plugin_file.read())
             return plugin_dict
-        raise Exception("No available plug-ins found. FACT backend might be down!")
+        raise Exception('No available plug-ins found. FACT backend might be down!')
 
     def get_binary_and_filename(self, uid):
         return self._request_response_listener(uid, 'raw_download_task', 'raw_download_task_resp')
 
+    def peek_in_binary(self, uid: str, offset: int, length: int) -> bytes:
+        ''' Return part of the content from file `uid` from offset `offset` with length `length`. '''
```
Comment (rhelmke) [inline src/intercom/front_end_binding.py:43]:
```suggestion
```

1. I think this comment is not required, as the code is self-explanatory :-)
2. codacy complains here about the whitespaced surrounding of this docstring

File: src/intercom/front_end_binding.py [hunk 2]
```diff
@@ -55,7 +59,7 @@ def get_binary_search_result(self, request_id):
     def _request_response_listener(self, input_data, request_connection, response_connection):
         serialized_request = pickle.dumps(input_data)
         request_id = generate_task_id(input_data)
-        self.connections[request_connection]['fs'].put(serialized_request, filename="{}".format(request_id))
+        self.connections[request_connection]['fs'].put(serialized_request, filename='{}'.format(request_id))
```
Comment (rhelmke) [inline src/intercom/front_end_binding.py:62]:
```suggestion
        self.connections[request_connection]['fs'].put(serialized_request, filename=request_id)
```

`generate_task_id` always returns a string :-)

File: src/web_interface/components/hex_highlighting.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            return highlighting_class
+    return None
+
+
+def _get_html_span(highlighting_class: str) -> str:
+    return f'<span class="hljs-{highlighting_class}">'  # reuse highlight.js classes for highlighting
+
+
+def _chr_to_hex(char: int) -> str:
+    ''' l => 6C '''
```
Comment (rhelmke) [inline src/web_interface/components/hex_highlighting.py:73]:
```suggestion
```

I do not fully understand the docstring here. Does it mean "the length of the output str is 6 characters long"? Also, Codacy does complain about the whitespaces here as well :-).

**Review by rhelmke — CHANGES_REQUESTED** [review]

File: src/web_interface/components/hex_highlighting.py (hunk 3 above)
Comment (jstucke) [reply src/web_interface/components/hex_highlighting.py:73]: It was just meant as an example: b"l" produces the output "6C". As that is also pretty self-explanatory I will remove it :sweat_smile:

File: src/intercom/front_end_binding.py (hunk 2 above)
Comment (jstucke) [reply src/intercom/front_end_binding.py:62]: Pre-commit fixed that one, I didn't even look closer at it :sweat_smile:  => fixed

File: src/intercom/front_end_binding.py (hunk 1 above)
Comment (jstucke) [reply src/intercom/front_end_binding.py:43]: I think you are right => removed the comment

**Review by rhelmke — APPROVED** [review]

**Review by rhelmke — APPROVED** [review]

**Pull request merged**